beacon-chain/sync/broadcast_bls_changes.go

```diff
@@ -26,11 +32,58 @@ func (s *Service) broadcastBLSChanges(currSlot types.Slot) {
 		return
 	}
 	source := rand.NewGenerator()
-	broadcastChanges := make([]*ethpb.SignedBLSToExecutionChange, len(changes))
-	for i := 0; i < len(changes); i++ {
+	length := len(changes)
+	broadcastChanges := make([]*ethpb.SignedBLSToExecutionChange, length)
+	for i := 0; i < length; i++ {
 		idx := source.Intn(len(changes))
 		broadcastChanges[i] = changes[idx]
 		changes = append(changes[:idx], changes[idx+1:]...)
 	}
-	s.cfg.p2p.BroadcastBLSChanges(s.ctx, broadcastChanges)
+
+	go s.rateBLSChanges(s.ctx, broadcastChanges)
+}
+
+func (s *Service) broadcastBLSBatch(ctx context.Context, ptr *[]*ethpb.SignedBLSToExecutionChange) {
+	changes := *ptr
+	limit := broadcastBLSChangesRateLimit
+	if len(changes) < broadcastBLSChangesRateLimit {
+		limit = len(changes)
+	}
+	st, err := s.cfg.chain.HeadState(ctx)
+	if err != nil {
+		log.WithError(err).Error("could not get head state")
+		return
+	}
+	for _, ch := range changes[:limit] {
+		if ch != nil {
+			_, err := blocks.ValidateBLSToExecutionChange(st, ch)
+			if err != nil {
+				log.WithError(err).Error("could not validate BLS to execution change")
+				continue
+			}
+			if err := s.cfg.p2p.Broadcast(ctx, ch); err != nil {
+				log.WithError(err).Error("could not broadcast BLS to execution changes.")
+			}
+		}
+	}
+	changes = changes[limit:]
+}
+
+func (s *Service) rateBLSChanges(ctx context.Context, changes []*ethpb.SignedBLSToExecutionChange) {
```

Comment:
Seems like duplicate logic

Reply:
Yes it is, the design decisions were pointed in the PR description, for the full discussion see here:
https://offchainlabs.slack.com/archives/C0477FMTB3K/p1675097230721769